### PR TITLE
Update HLS installation method

### PR DESCRIPTION
**Pull request by hasufell**
@jneira 

https://github.com/haskell/haskell-language-server/issues/2739

### CODE REVIEW

File: src/extension.ts
Comment (hasufell) [inline src/extension.ts:243]: Somehow this doesn't work... I tried adding prints and somewhere here it seems to silently crash or whatever. I don't have the patience to investigate this js nonsense.

Installation itself seems to work, but now we want HLS to see the rest of the binaries where `haskell-language-wrapper` resides.

**Comment by jneira** [comment]
sorry, don't have time to review the pr, hope another maintainer could do it

File: src/extension.ts
Comment (pepeiborra) [reply src/extension.ts:243]: What other binaries? The extension should only need the path to the HLS-wrapper executable and the path to the HLS executable.

File: src/hlsBinaries.ts
Comment (wz1000) [inline src/hlsBinaries.ts:281]: We could also check if the user already has GHCup and use that if it exists.

File: src/extension.ts
Comment (pepeiborra) [inline src/extension.ts:197]: After getting the project version, shouldn't we use it to verify that we have installed a matching HLS executable?

File: src/hlsBinaries.ts
Comment (fendor) [inline src/hlsBinaries.ts:167]:
```suggestion
                "No version of HLS installed or found and updateBehaviour set to 'never-check' giving up..."
```

Missing a space

File: src/hlsBinaries.ts
Comment (fendor) [inline src/hlsBinaries.ts:175]:
```suggestion
                throw new Error('No version of HLS installed or found and installation was denied, giving up...');
```

Again, space missing.

File: src/hlsBinaries.ts
Comment (pepeiborra) [reply src/hlsBinaries.ts:281]: But then we need to upgrade it before using, etc.

**Review by fendor — COMMENTED** [review]
Just some comments for now

File: src/extension.ts
Comment (fendor) [reply src/extension.ts:243]:
```
stack trace: Error: TypeError: Cannot assign to read only property 'PATH' of object
	at Object.set (/usr/lib/code/out/vs/workbench/services/extensions/node/extensionHostProcess.js:97:44518)
	at /home/hugin/Documents/haskell/vscode-haskell/dist/extension.js:212:39
	at Generator.next (<anonymous>)
	at fulfilled (/home/hugin/Documents/haskell/vscode-haskell/dist/extension.js:11:58)
```
I can help out with the extension a bit

Comment (fendor) [reply src/extension.ts:243]:
```
stack trace: Error: TypeError: Cannot assign to read only property 'PATH' of object
	at Object.set (/usr/lib/code/out/vs/workbench/services/extensions/node/extensionHostProcess.js:97:44518)
	at /home/hugin/Documents/haskell/vscode-haskell/dist/extension.js:212:39
	at Generator.next (<anonymous>)
	at fulfilled (/home/hugin/Documents/haskell/vscode-haskell/dist/extension.js:11:58)
```
I can help out with the extension a bit

Comment (fendor) [reply src/extension.ts:243]:
```suggestion
  let serverEnvironment: IEnvVars = workspace.getConfiguration('haskell', uri).serverEnvironment;
  if (addInternalServerPath != undefined) {
      const pathSep = process.platform === 'win32' ? ';' : ':';
      const PATH = process.env.PATH!.split(pathSep);
      PATH.unshift(addInternalServerPath);
      serverEnvironment = {
        PATH: PATH.join(pathSep),
        ... serverEnvironment
      };
  }
```

This seems to work

Comment (fendor) [reply src/extension.ts:243]: BTW, you can see errors of your extension by looking at "Debug Console" of the original VSCode Window (the one that contains the extension's code)

File: src/hlsBinaries.ts
Comment (hasufell) [reply src/hlsBinaries.ts:281]: Yes, I want to avoid messing with user-installed ghcup.

File: src/extension.ts
Comment (hasufell) [reply src/extension.ts:197]: Yeah

File: src/extension.ts
Comment (hasufell) [reply src/extension.ts:243]: > What other binaries? The extension should only need the path to the HLS-wrapper executable and the path to the HLS executable.

Yes, but the path to the hls-wrapper is not in PATH. It didn't work here.

Comment (hasufell) [reply src/extension.ts:243]: > This seems to work

Nice, that works indeed.

File: src/extension.ts
Comment (hasufell) [reply src/extension.ts:197]: Screenshot_2022-02-28_12-28-19

**Comment by hasufell** [comment]
I'm quite pleased. We now:

1. have our completely independent `<storagePath>/.ghcup` prefix where we manage HLS versions
2. we can easily remove old versions, update etc.
3. we could theoretically use this to pre-install cabal/ghc/stack as well

For the future, the extension could also have an option for the user to use the system-installed `ghcup`. But this really needs to be behind a config option imo.

**Comment by fendor** [comment]
> For the future, the extension could also have an option for the user to use the system-installed ghcup. But this really needs to be behind a config option imo.

I am interested in that, and I agree, it should be hidden behind a config option.

**Comment by hasufell** [comment]
Some tests are failing, probably because we started messing with `process.env`:

```
2 failing

  1) Extension Test Suite
       Extension log should have server output:
     AssertionError [ERR_ASSERTION]: Extension log file has no hls output
  	at /home/runner/work/vscode-haskell/vscode-haskell/out/test/suite/extension.test.js:136:16
  	at Generator.next (<anonymous>)
  	at fulfilled (out/test/suite/extension.test.js:5:58)


  2) Extension Test Suite
       Server should inherit environment variables defined in the settings:

      AssertionError [ERR_ASSERTION]: Server did not inherit XDG_CACHE_DIR from environment variables set in the settings
      + expected - actual

      -false
      +true
      
  	at /home/runner/work/vscode-haskell/vscode-haskell/out/test/suite/extension.test.js:140:16
  	at Generator.next (<anonymous>)
  	at fulfilled (out/test/suite/extension.test.js:5:58)
```

Ideas?

**Comment by hasufell** [comment]
I'm assuming the extension used to try to find older HLS releases if the current project GHC is too new?

**Comment by michaelpj** [comment]
> I'm assuming the extension used to try to find older HLS releases if the current project GHC is too old?

Yes, and this was a highly requested feature that we should definitely not break.

File: src/hlsBinaries.ts
Comment (hasufell) [inline src/hlsBinaries.ts:500]: So this is obviously a temporary url. It contains the HLS-GHC mapping: https://gist.githubusercontent.com/hasufell/dd84df5f81a3a7e6e6fad8f122dba429/raw/73efc1078555d971076d3ccf31154f10ed683a82/hls-metadata.yaml

From this we figure out the latest HLS version that supports a given GHC.

The idea is that ghcup will generate this file on new releases, see https://github.com/haskell/ghcup-metadata/pull/22

**Comment by hasufell** [comment]
> > I'm assuming the extension used to try to find older HLS releases if the current project GHC is too old?
> 
> Yes, and this was a highly requested feature that we should definitely not break.

This is fixed in https://github.com/haskell/vscode-haskell/pull/533/commits/cda68baa9f03f4715eb5aa9a70b409e6f6640353

Example:
Screenshot_2022-03-01_00-58-33

**Comment by hasufell** [comment]
I don't understand why the tests are failing. I hope someone else can pick that up, because I'm running out of free time for this.

**Comment by hasufell** [comment]
Well, I'm giving up on the test suite. It works locally. That's enough for me.

**Comment by fendor** [comment]
Ill take a look

**Comment by hasufell** [comment]
Alright. The tests are broken. The installation of HLS works though, including on windows, I just tested it manually.

Feel free to merge this PR or adjust it.

File: src/hlsBinaries.ts
Comment (fendor) [reply src/hlsBinaries.ts:583]: I was suggesting this should be hosted GHCup, but since this actually purely HLS metadata, we should host it somewhere assocaited with HLS... Maybe as a file in HLS directly?

File: .github/workflows/test.yml
Comment (fendor) [inline .github/workflows/test.yml:47]: Add comment to explain we are doing this to avoid test-timeouts

File: .github/workflows/test.yml
Comment (fendor) [inline .github/workflows/test.yml:73]: Somehow document this is necessary because the tests don't clean up after themselves.

**Review by fendor — CHANGES_REQUESTED** [review]

**Comment by fendor** [comment]
This PR is hard to review, and probably also hard to split into sensible commits... 

Maybe we should just merge it and deal with fallouts later?

**Comment by hasufell** [comment]
> This PR is hard to review, and probably also hard to split into sensible commits...
> 
> Maybe we should just merge it and deal with fallouts later?

That's why code owners should review.

**Comment by fendor** [comment]
We don't have explicit code-owners for this project...

Afaict, the only real recent "code-owner" (rather project owner) is jneira and I believe they don't have time recently to review such a big PR in depth. (So, I don't want to stress jneira out)
Previously, the download logic was written by bubba, but I haven't seen bubba in quite a while.

File: src/hlsBinaries.ts
Comment (hasufell) [reply src/hlsBinaries.ts:583]: > I was suggesting this should be hosted GHCup, but since this actually purely HLS metadata, we should host it somewhere assocaited with HLS... Maybe as a file in HLS directly?

The metadata only changes when ghcup changes metadata, so I'd say it should be in ghcup-metadata repo.

Comment (fendor) [reply src/hlsBinaries.ts:583]: It is totally legitimate in the ghcup-metadata repo, so yes, let's go for it!

**Comment by hasufell** [comment]
> Afaict, the only real recent "code-owner" (rather project owner) is jneira and I believe they don't have time recently to review such a big PR in depth.

Then those maintainers should look for their successors. Since I don't use VSCode, I'm not the right person. Otherwise we should consider marking this project as unmaintained. Just keeping it on life-support with casual PRs will probably make maintenance even harder.

**Comment by fendor** [comment]
I don't think we need to look for a successor, sometimes maintainers are just pressed for time.
Anyway, I think it is in general a good idea to have multiple maintainers, and I am happy to take more responsibility in this project (and I *am* a heavy vscode user).

**Comment by jneira** [comment]
@fendor thanks for you head up and your understanding

I never was the project or code owner. This project, which started before i was around, needed some love and it turns out i've spent some time on it.  I did not have any typescript or vscode api knowledge before that. So i was not the succesor of nobody and i dont think i have to look for a succesor for something never was "mine".

Technically the entire [haskell ide team](https://github.com/orgs/haskell/teams/haskell-ide) is administrator of the project and can do any change in the project.

**Comment by hasufell** [comment]
@jneira you're the code owner of the modules that needed changes (according to git log) and the no. 1 contributor to this project: https://github.com/haskell/vscode-haskell/graphs/contributors

I would have appreciated a review here, because I don't want to break VSCode user experience and have had a lot of struggle with this codebase I'm totally unfamiliar with. But it appears you've disengaged from all PRs I've raised so far wrt dynamic bindists  and don't want to get involved with it.

**Comment by jneira** [comment]
> @jneira you're the code owner of the modules that needed changes (according to git log) and the no. 1 contributor to this project: https://github.com/haskell/vscode-haskell/graphs/contributors
> 

There are lies, damn lies and statistics. 😆 
 I have not done any substancial change in the existing codebase. Nothing comparable with this pr for example. Most changes ere administrative ones (and i bet most of those lines are from the damn `package-lock.json`)
The more substantial change was add the test suite, driven by googling "how to do x in typescript", and it seems it has caused more problems than benefits 🤷 

> I would have appreciated a review here, because I don't want to break VSCode user experience and have had a lot of struggle with this codebase I'm totally unfamiliar with. But it appears you've disengaged from all PRs I've raised so far wrt dynamic bindists and don't want to get involved with it.

As i reirated many times I don't have as much time as before for no pr or issue in hls or vscode independently of their authors, so i can assure you there is no pesonal reason for that.

File: .github/workflows/test.yml
Comment (fendor) [inline .github/workflows/test.yml:44]:
```suggestion
      # per HLS version one HLS binary per GHC version. 
```

File: test/suite/extension.test.ts
Comment (fendor) [inline test/suite/extension.test.ts:127]:
```suggestion
```

File: test/suite/extension.test.ts
Comment (fendor) [inline test/suite/extension.test.ts:150]:
```suggestion
```

File: test/suite/extension.test.ts
Comment (fendor) [inline test/suite/extension.test.ts:182]:
```suggestion
```

**Review by fendor — CHANGES_REQUESTED** [review]

**Review by fendor — APPROVED** [review]
LGTM

**Comment by hasufell** [comment]
Here was my attempt at adding a `systemGHCup` option:

```diff
From 78fdd4dadf0430635d00d6439db33b093315bb97 Mon Sep 17 00:00:00 2001
From: Julian Ospald <hasufell@posteo.de>
Date: Wed, 9 Mar 2022 21:07:24 +0100
Subject: [PATCH] Allow to use system ghcup

---
 package.json       |  6 ++++++
 src/hlsBinaries.ts | 13 +++++++++----
 2 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/package.json b/package.json
index cbd2b62..b8c842e 100644
--- a/package.json
+++ b/package.json
@@ -173,6 +173,12 @@
           "default": false,
           "description": "Whether to ignore haskell-language-server on PATH"
         },
+        "haskell.useSystemGHCup": {
+          "scope": "resource",
+          "type": "boolean",
+          "default": "false",
+          "description": "Whether to use the system ghcup or an internal one for installing HLS."
+        },
         "haskell.checkProject": {
           "scope": "resource",
           "type": "boolean",
diff --git a/src/hlsBinaries.ts b/src/hlsBinaries.ts
index 82414ad..28e0542 100644
--- a/src/hlsBinaries.ts
+++ b/src/hlsBinaries.ts
@@ -289,10 +289,15 @@ async function callGHCup(
   cancellable?: boolean
 ): Promise<string> {
   const storagePath: string = await getStoragePath(context);
-  const ghcup = path.join(storagePath, `ghcup${exeExt}`);
-  return await callAsync(ghcup, ['--no-verbose'].concat(args), storagePath, logger, title, cancellable, {
-    GHCUP_INSTALL_BASE_PREFIX: storagePath,
-  });
+  const systemGHCup = workspace.getConfiguration('haskell').get('useSystemGHCup') as boolean;
+  if (systemGHCup) {
+	  return await callAsync('ghcup', ['--no-verbose'].concat(args), storagePath, logger, title, cancellable);
+  } else {
+	  const ghcup = path.join(storagePath, `ghcup${exeExt}`);
+	  return await callAsync(ghcup, ['--no-verbose'].concat(args), storagePath, logger, title, cancellable, {
+		GHCUP_INSTALL_BASE_PREFIX: storagePath,
+	  });
+  }
 }
 
 async function getLatestSuitableHLS(
-- 
2.34.1

```

But after hacking on this extension for a while, everything stopped compiling altogether with obscure npm errors. `git clean -fdx` and trying from scratch didn't work, so I cannot compile this project anymore.

**Comment by hasufell** [comment]
I added a `useSystemGHCup` option... however, because of this

```ts
    // Try and find local installations first
    serverExecutable = findManualExecutable(logger, uri, folder) ?? findLocalServer(context, logger, uri, folder);
    if (serverExecutable === null) {
      // If not, then try to download haskell-language-server binaries if it's selected
      await downloadGHCup(context, logger);
      serverExecutable = await downloadHaskellLanguageServer(context, logger, currentWorkingDir);
      await validateHLSToolchain(serverExecutable, currentWorkingDir, logger);
      addInternalServerPath = path.dirname(serverExecutable);
      if (!serverExecutable) {
        return;
      }
    }
```

this feature is a little odd... if we find anything in PATH, it will be picked, even if it's an incompatible HLS (that doesn't support the project GHC). So to get vscode to manage HLS with system ghcup and install all required versions, you also have to set `ignorePATH` to `true`.

**Comment by hasufell** [comment]
I pushed my config idea to this PR. Please test/review and feel free to discard it. If it's accepted, it'll probably require some changelog or user notification. Up to you.

**Comment by fendor** [comment]
I like the re-organisation of options and am completely in favour. There is one use-case not covered by this, afaict:
* Let VSCode manage HLS and ignore anything that exists previously

I am not sure whether there is a use-case when this combination is relevant / desired (maybe if it is a per workspace setting?)

A possible work-around would be to turn `manageHLS` into a dropdown with options such as `manual` / `automatic` / `isolationist` (where automatic is the default and what you described above).

However, iirc, you argued that PATH manipulation should not happen in the Extension and is an orthogonal concern that can be handled by either other extensions or other tools, right? 

@soiamsoNG does what's described [in the comment above](https://github.com/haskell/vscode-haskell/pull/533#issuecomment-1063396466) work for you?

**Comment by hasufell** [comment]
> Let VSCode manage HLS and ignore anything that exists previously

I don't really understand this use case either.

1. ignoring system ghcup provided HLS doesn't make much sense... because it's the same as the extension would install
2. ignoring HLS installed by cabal can be done by e.g. removing `~/.cabal/bin` from PATH or uninstalling it
3. ignoring HLS installed system-wide, e.g. in `/usr/bin` would be a tad harder, because you can't remove it from PATH easily... I somewhat doubt that this is a common setup though

**Comment by fendor** [comment]
Regarding 2., I did run into this use case once, I was testing a patch I was submitting to HLS, but it was slightly buggy and I was debugging the issue.
Then I had to switch gears and was working on something else, but suddenly was using my borked HLS installation instead of the good one. I did not want to remove the HLS installation, as rebuilding can take some time (e.g. when you switch branch, build something, then switch branch back to what you have been working on).
In this very rare, special case, I would have liked the option to disable discovering HLS versions from the PATH for certain workspaces (or the other way around, allowing discovering HLS from the PATH for certain workspaces).

Anyway, it is a rather contrived example, so I am fine with the change as is, but let's see what @soiamsoNG's use-case is.

**Comment by hasufell** [comment]
> I did not want to remove the HLS installation, as rebuilding can take some time (e.g. when you switch branch, build something, then switch branch back to what you have been working on).

With cabal v2 the binary you install is a symlink. Removing it will not change the cabal store and re-installing it will take 2s.

**Comment by soiamsoNG** [comment]
for haskell.ignorePath similar as the case you mention @fendor https://github.com/haskell/vscode-haskell/pull/533#issuecomment-1064956508

BTW, for this PR, I check https://github.com/haskell/vscode-haskell/pull/533/commits/6374d03e09f9495467cf7f9a41448bb8e419cafa  the haskell.releasesUrl part, which have no effect in the internal ghcup downloading, I recommend expose the internal [ghcup mirror](https://www.haskell.org/ghcup/guide/#mirrors)  setting. maybe need make one more option or reuse haskell.releasesUrl

It will be more easy for setup inside devcontainer. In my case can't get the default link use by ghcup

**Comment by hasufell** [comment]
> for haskell.ignorePath similar as the case you mention @fendor https://github.com/haskell/vscode-haskell/pull/533#issuecomment-1064956508

I still think it's not a generally interesting setting for most users and rather confusing. For development, you have the options to fix your PATH or temporarily delete experimental binaries before you launch HLS.

> BTW, for this PR, I check https://github.com/haskell/vscode-haskell/commit/6374d03e09f9495467cf7f9a41448bb8e419cafa the haskell.releasesUrl part, which have no effect in the internal ghcup downloading, I recommend expose the internal [ghcup mirror](https://www.haskell.org/ghcup/guide/#mirrors) setting. maybe need make one more option or reuse haskell.releasesUrl

Yes, I'm ware that e.g. users from china need to use mirrors, so this seems like an important setting.

**Comment by hasufell** [comment]
An alternative is that we just ignore PATH altogether if `haskell.manageHLS` is set to `true`. I think this might in fact be better. Just because we have the correct HLS wrapper version in PATH doesn't mean the appropriate `haskell-language-server-<ghcver>` is there as well... potentially confusing the extension. That can be fixed too with file-existence checks, but this kind of complexity is exactly what I tried to avoid.

**Comment by hasufell** [comment]
Ok, I made a few changes:

1. `haskell.useSystemGHCup` starts with a null value and on first extension start it will prompt the user if they want to use system ghcup
2. `PATH` is now completely ignored as long as `haskell.manageHLS == true`
3. if `haskell.manageHLS == false`, then we check `haskell.serverExecutablePath` and fall back to checking PATH... no weird cross-lookup logic
4. I added `haskell.metadataURL` to support mirrors

Also made a couple of changes wrt installation and popup behavior. It should be much smoother now.

**Review by fendor — APPROVED** [review]
Re-approving since there have been a ton of changes

**Comment by hasufell** [comment]
hls metadata fixed for windows: https://github.com/haskell/ghcup-metadata/commit/9d2ebcb74d1c53183808577993ecd8dbe22e6aeb

**Pull request merged**
